tomviz/PipelineModel.cxx

```diff
@@ -0,0 +1,596 @@
+  limitations under the License.
+
+******************************************************************************/
+#include "PipelineModel.h"
+
+#include "ActiveObjects.h"
+#include "DataSource.h"
+#include "Module.h"
+#include "ModuleManager.h"
+#include "Operator.h"
+
+#include <QFileInfo>
+
+namespace tomviz {
+
+struct PipelineModel::Item
+{
+  Item(DataSource *source) : tag(DATASOURCE), s(source) { }
+  Item(Module *module) : tag(MODULE), m(module) { }
+  Item(Operator *op) : tag(OPERATOR), o(op) { }
+
+  DataSource* dataSource() { return tag == DATASOURCE ? s : nullptr; }
+  Module* module() { return tag == MODULE ? m : nullptr; }
+  Operator* op() { return tag == OPERATOR ? o : nullptr; }
+
+  enum { DATASOURCE, MODULE, OPERATOR } tag;
+  union
+  {
+    DataSource *s;
+    Module *m;
+    Operator *o;
+  };
+};
+
+class PipelineModel::TreeItem
+{
+public:
+  explicit TreeItem(const PipelineModel::Item &item,
+                    TreeItem *parent = nullptr);
+  ~TreeItem();
+
+  TreeItem* parent() { return m_parent; }
+  TreeItem* child(int index);
+  int childCount() const { return m_children.count(); }
+  QList<TreeItem*>& children() { return m_children; }
+  int childIndex() const;
+  bool appendChild(const PipelineModel::Item &item);
+  bool appendAndMoveChildren(const PipelineModel::Item &item);
+  bool insertChild(int position, const PipelineModel::Item &item);
+  bool removeChild(int position);
+  bool moveChildren(TreeItem *newParent);
+
+  bool remove(DataSource *source);
+  bool remove(Module *module);
+  bool remove(Operator *op);
+
+  bool hasOp(Operator *op);
+
+  TreeItem* find(Module *module);
+
+  DataSource* dataSource() { return m_item.dataSource(); }
+  Module* module() { return m_item.module(); }
+  Operator* op() { return m_item.op(); }
+
+private:
+  QList<TreeItem*> m_children;
+  PipelineModel::Item m_item;
+  TreeItem *m_parent;
+};
+
+PipelineModel::TreeItem::TreeItem(const PipelineModel::Item &i, TreeItem *p)
+  : m_item(i), m_parent(p)
+{
+}
+
+PipelineModel::TreeItem::~TreeItem()
+{
+  qDeleteAll(m_children);
+}
+
+PipelineModel::TreeItem* PipelineModel::TreeItem::child(int index)
+{
+  return m_children.value(index);
+}
+
+int PipelineModel::TreeItem::childIndex() const
+{
+  if (m_parent) {
+    return m_parent->m_children.indexOf(const_cast<TreeItem*>(this));
+  }
+  return 0;
+}
+
+bool PipelineModel::TreeItem::insertChild(int pos,
+                                          const PipelineModel::Item &item)
+{
+  if (pos < 0 || pos >= m_children.size()) {
+    return false;
+  }
+  TreeItem *treeItem = new TreeItem(item, this);
+  m_children.insert(pos, treeItem);
+  return true;
+}
+
+bool PipelineModel::TreeItem::appendChild(const PipelineModel::Item &item)
+{
+  TreeItem *treeItem = new TreeItem(item, this);
+  m_children.append(treeItem);
+  return true;
+}
+
+bool PipelineModel::TreeItem::appendAndMoveChildren(const PipelineModel::Item &item)
+{
+  auto treeItem = new TreeItem(item, this);
```

Comment:
We're allowed to use auto now?  (Not that I mind, I've been wanting to for a while).